pnpm/test/hooks.ts

```diff
@@ -223,3 +224,75 @@ test('custom fetcher can call default fetcher', async () => {
 
   expect(args.opts).toBeDefined()
 })
+
+test('adding or changing pnpmfile should change pnpmfileChecksum and module structure', async () => {
+  const project = prepare({
+    dependencies: {
+      '@pnpm.e2e/pkg-with-good-optional': '1.0.0',
+    },
+  })
+
+  await execPnpm(['install'])
+
+  const lockfile0 = project.readLockfile()
+  expect(lockfile0.pnpmfileChecksum).toBeUndefined()
+  expect(lockfile0.packages).toHaveProperty(['/@pnpm.e2e/pkg-with-good-optional@1.0.0'])
+  expect(lockfile0.packages).toHaveProperty(['/is-positive@1.0.0'])
+
+  const pnpmfile1 = `
+    function readPackage (pkg) {
+      if (pkg.optionalDependencies) {
+        pkg.optionalDependencies = {} // 'delete pkg.optionalDependencies' doesn't work here
```

Comment:
@zkochan If you replace this line with `delete pkg.optionalDependencies`, the test would fail.

Reply:
I have found the issue. I'll submit another PR to fix it.

Reply:
Here's the problem:

https://github.com/pnpm/pnpm/blob/3ded840001b76c828cd914fb5bad74585274f7ba/pkg-manager/resolve-dependencies/src/resolveDependencies.ts#L1255

`optionalDependencies` from the lockfile is not overwritten.